File: kotlin/vci-client/src/test/java/io/mosip/vciclient/credentialRequest/types/LdpVcCredentialRequestTest.kt
package io.mosip.vciclient.credentialRequest.types

import io.mosip.vciclient.constants.CredentialFormat
import io.mosip.vciclient.credentialRequest.util.ValidatorResult
import io.mosip.vciclient.proof.jwt.JWTProof
import io.mosip.vciclient.dto.IssuerMetaData
import okhttp3.Request
import okhttp3.internal.http2.Header
import okhttp3.internal.toHeaderList
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import java.net.URI

class LdpVcCredentialRequestTest {
    @Test
    fun `should create JSON in expected format`() {
        val credentialEndpoint = "https://credentialendpoint/"

        val ldpVcRequest: Request = LdpVcCredentialRequest(
            "accessToken",
            IssuerMetaData(
                "/credentialAudience",
                credentialEndpoint,
                30000,
                credentialType = arrayOf("VerifiableCredential"),
                credentialFormat = CredentialFormat.LDP_VC
            ), JWTProof("headerEncoded.payloadEncoded.signature")
        ).constructRequest()

        assertTrue(
            ldpVcRequest.headers.toHeaderList().containsAll(
                listOf(
                    Header("Authorization", "Bearer accessToken"),
                    Header("Content-Type", "application/json")
                )
            )
        )
        assertEquals(URI(credentialEndpoint), ldpVcRequest.url.toUri())
        assertEquals("POST", ldpVcRequest.method)
    }

    @Test
    fun `should return isValid as true when required issuerMetadata details are available`() {

        val ldpVcRequest: LdpVcCredentialRequest = LdpVcCredentialRequest(
            "accessToken",
            IssuerMetaData(
                "/credentialAudience",
                "https://credentialendpoint/",
                30000,
                credentialType = arrayOf("VerifiableCredential"),
                credentialFormat = CredentialFormat.LDP_VC
            ), JWTProof("headerEncoded.payloadEncoded.signature")
        )

        val validatorResult: ValidatorResult = ldpVcRequest.validateIssuerMetaData()

        assertTrue(validatorResult.isValid)
        assertTrue(validatorResult.invalidFields.isEmpty())
    }

    @Test
    fun `should return validator result with isValid as false & invalidFields when required issuerMetadata details are not available`() {

        val ldpVcRequest = LdpVcCredentialRequest(
            "accessToken",
            IssuerMetaData(
                "/credentialAudience",
                "https://credentialendpoint/",
                30000,
                credentialFormat = CredentialFormat.LDP_VC
            ), JWTProof("headerEncoded.payloadEncoded.signature")
        )

        val validatorResult: ValidatorResult = ldpVcRequest.validateIssuerMetaData()

        assertFalse(validatorResult.isValid)
        assertTrue(validatorResult.invalidFields.containsAll(listOf("credentialType")))
    }
}

File: js/test/credentialRequest/types/LdpVcCredentialRequest.test.js
const LdpVcCredentialRequest = require("../../../src/credentialRequest/types/LdpVcCredentialRequest");
const JWTProof = require("../../../src/proof/jwt/JWTProof");
const IssuerMetaData = require('../../../src/dto/issuerMetaData');

test("should create JSON in expected format", () => {
  const credentialEndpoint = "https://domain.net/credential";
  const credentialAudience = "https://domain.net";
  const downloadTimeoutInMillSeconds = 3000;
  const credentialType = ["VerifiableCredential"];
  const proof = new JWTProof("eyJhbdFQwRnFWZjIcQku5nL1E_d-uifyj84U3XU8TA");
  const credentialFormat = "ldp_vc";
  const issuerMetaData = new IssuerMetaData(
    credentialAudience,
    credentialEndpoint,
    downloadTimeoutInMillSeconds,
    credentialType,
    credentialFormat,
  );

  const expectedRequestHeader = {
    Authorization: "Bearer accessToken",
    "Content-Type": "application/json",
  };
  const expectedRequestBody =
    '{"format":"ldp_vc","credential_definition":{"type":["VerifiableCredential"],"@context":["https://www.w3.org/2018/credentials/v1"]},"proof":{"proof_type":"jwt","jwt":"eyJhbdFQwRnFWZjIcQku5nL1E_d-uifyj84U3XU8TA"}}';

  const ldpVcRequest = new LdpVcCredentialRequest(
    "accessToken",
    issuerMetaData,
    proof
  ).constructRequest();

  expect(ldpVcRequest.requestHeader).toEqual(expectedRequestHeader);
  expect(ldpVcRequest.requestBody).toEqual(expectedRequestBody);
});
